atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java

```diff
@@ -710,9 +710,12 @@ private boolean isEntryConsistent(final long newEntryPosition) {
       } while (index > 0 && lastEntry != null && lastEntry.type() != ZeebeEntry.class);
     }
 
-    return lastEntry == null
-        || lastEntry.type() != ZeebeEntry.class
-        || newEntryPosition > ((ZeebeEntry) lastEntry.entry()).highestPosition();
+    long lastPosition = -1;
+    if (lastEntry != null && lastEntry.type() == ZeebeEntry.class) {
+      lastPosition = ((ZeebeEntry) lastEntry.entry()).highestPosition();
+    }
+
+    listener.validatePositions(lastPosition, entry);
```

Comment:
I switched to the exception so we can build it in the validating function and give it more information about the failure. We can say "we tried appending entry X and there was a gap" but we can't say there was a gap between X and X+2. The gap might be between the last entry's highest position and the current entry's lowest or it might be that the a gap is detected but the current entry has the correct positions but in the wrong order. These are different situations and a more descriptive log can provide useful context. But maybe we can do this in a better way

Reply:
You could also have a validation result object which contains an valid (true/false) property and an description of the error. I think this is normally the way.

Reply:
Example https://medium.com/@aleksanderkolata/java-tips-01-handle-validation-in-an-elegant-manner-using-chain-of-responsibility-design-pattern-ad3dcc22479e

Reply:
Sounds good, I'll do that